packages/ui-voip/src/providers/useMediaSession.ts

```diff
@@ -0,0 +1,291 @@
+	hidden: false,
+};
+
+type MediaSession = SessionInfo & {
+	toggleMute: () => void;
+	toggleHold: () => void;
+
+	toggleWidget: (peerInfo?: PeerInfo) => void;
+	selectPeer: (peerInfo: PeerInfo) => void;
+
+	endCall: () => void;
+	startCall: (id: string, kind: 'user' | 'sip') => Promise<void>;
+	acceptCall: () => Promise<void>;
+
+	changeDevice: (deviceId: string) => Promise<void>;
+
+	forwardCall: (type: 'user' | 'sip', id: string) => void;
+	sendTone: (tone: string) => void;
+};
+
+const deriveWidgetStateFromCallState = (callState: CallState, callRole: CallRole): State | undefined => {
+	switch (callState) {
+		case 'active':
+		case 'accepted':
+		case 'renegotiating':
+			return 'ongoing';
+		case 'none':
+		case 'ringing':
+			return callRole === 'callee' ? 'ringing' : 'calling';
+	}
+};
+
+const deriveConnectionStateFromCallState = (callState: CallState): ConnectionState => {
+	switch (callState) {
+		case 'renegotiating':
+			return 'RECONNECTING';
+		case 'ringing':
+		case 'active':
+			return 'CONNECTED';
+		case 'none':
+		case 'accepted':
+		default:
+			return 'CONNECTING';
+	}
+};
+
+const reducer = (
+	reducerState: SessionInfo,
+	action: {
+		type: 'toggleWidget' | 'selectPeer' | 'instance_updated' | 'reset' | 'mute' | 'hold';
+		payload?: Partial<SessionInfo>;
+	},
+): SessionInfo => {
+	if (action.type === 'mute') {
+		return {
+			...reducerState,
+			muted: action.payload?.muted ?? reducerState.muted,
+		};
+	}
+
+	if (action.type === 'hold') {
+		return {
+			...reducerState,
+			held: action.payload?.held ?? reducerState.held,
+		};
+	}
+
+	if (action.type === 'toggleWidget') {
+		if (reducerState.state === 'closed') {
+			return { ...reducerState, state: 'new', peerInfo: action.payload?.peerInfo };
+		}
+
+		if (reducerState.state === 'new') {
+			return { ...reducerState, state: 'closed' };
+		}
+	}
+
+	if (action.type === 'instance_updated') {
+		return { ...reducerState, ...action.payload } as SessionInfo;
+	}
+
+	if (action.type === 'selectPeer') {
+		if (reducerState.state !== 'new') {
+			return reducerState;
+		}
+
+		return { ...reducerState, peerInfo: action.payload?.peerInfo };
+	}
+
+	if (action.type === 'reset') {
+		return defaultSessionInfo;
+	}
+
+	return reducerState;
+};
+
+export const useMediaSession = (instance?: MediaSignalingSession): MediaSession => {
+	const [mediaSession, dispatch] = useReducer<typeof reducer>(reducer, defaultSessionInfo);
+
+	const getAvatarUrl = useUserAvatarPath();
+
+	useEffect(() => {
+		if (!instance) {
+			dispatch({ type: 'reset' });
+			return;
+		}
+
+		const updateSessionState = () => {
+			const mainCall = instance.getMainCall();
+
+			if (!mainCall) {
+				dispatch({ type: 'reset' });
+				return;
+			}
+
+			const { contact, state: callState, role, muted, held, hidden } = mainCall;
+			const state = deriveWidgetStateFromCallState(callState, role);
+			const connectionState = deriveConnectionStateFromCallState(callState);
+
+			if (contact.type === 'sip') {
```

Comment:
for the client, the `contact.type` attribute only identifies the type of value present in the `contact.id` attribute. You can (and should) still load the value of any other attribute that you receive, specially displayName.

Reply:
I'll keep external calls just with the number for now as we discussed.